### DISCUSSION TITLE
Godaddy Integration

### DISCUSSION THREAD

**Discussion by vigouredelaruse** [polymind-inc/acmebot#280] [Ideas]
godaddy has this api for managing domains https://developer.godaddy.com/doc/endpoint/domains

you cannot pull or push a zonefile using this api but you can discover the zones by iterating over the list domain apis

i want to use godaddy as the dns provider - i forked the project and started looking into the feasibility and right now i don't see why integrating godaddy as a dns provider could not be acconplished - what's the most efficient use of everyone's time

1) fork the code finish the godaddy integration myself and maintain the fork separately
2) wait for keyvault-acmebot to support godaddy either through its own efforts or using my code

extra comfort - here's php code updating txt records using the godaddy domains api
https://domainnamewire.com/2019/07/25/tutorial-how-to-update-dns-records-using-godaddy-api/

please advise

**Comment by shibayan** [comment]
Looking at the REST API provided by GoDaddy, I think I can implement a DNS provider without any problem.

I don't use GoDaddy, so I can't implement it. In most cases, the DNS Provider is implemented by a contributor.

**Reply by vigouredelaruse** [reply]
well all right i have a fork from yesterday - let me get it working in the fork then we can discuss how to integrate it into the main branch. agreed?

**Reply by shibayan** [reply]
If it works fine, you can make a pull request and I will review it and merge it.

If it doesn't work, I can't merge it, and I can't identify the cause.

**Reply by vigouredelaruse** [reply]
perhaps you could help me out quickly with this - 

what is the intent of providing a collection of values to be applied to TXT here? typically TXT record challenges i've seen usually have an equation or a single constant 

if there's more than one what should the delimiter be?

        public async Task CreateTxtRecordAsync(DnsZone zone, string relativeRecordName, IEnumerable<string> values)
        {
            var urlTemplate = "/domains/{zone}/records/{type}/{name}";
            var url = String.Format(urlTemplate, zone.Name, GodaddyConstants.DNS_TXT_RECORD, relativeRecordName);
            var txtRecord = new GoDaddyDnsRecord()
            {
                name = relativeRecordName,
                data = string.Join(" ", values) // how to join this
            };

            await _httpClient.PutAsJsonAsync<GoDaddyDnsRecord>(url, txtRecord);
        }

**Reply by shibayan** [reply]
Since the DNS can have multiple values, set only the amount needed by ACME. I'm not sure how this is provided by GoDaddy's API.

**Reply by vigouredelaruse** [reply]
for now i will make sure txt record processing supports the structured and unstructured text scenarios discussed here

https://en.wikipedia.org/wiki/TXT_record#:~:text=A%20TXT%20record%20(short%20for,center%2C%20or%20other%20accounting%20information. and here https://tools.ietf.org/html/rfc1464
As unstructured text, organisations can use the TXT string in any way they define, for example:

example.com.   IN   TXT   "This domain name is reserved for use in documentation"

**Comment by jgross-OBA** [comment]
Hi, I am also interested in having GoDaddy available as a provider so certainly would be willing to work with you on getting this to work.

Looking at the code for the Goggle DNS Provider it doesn't have the join logic for the TXT record so that may be a better example. I know in testing with POSH-ACME the GoDaddy DNS code only deals with a single TXT record:

public async Task CreateTxtRecordAsync(DnsZone zone, string relativeRecordName, IEnumerable<string> values)
        {
            var recordName = $"{relativeRecordName}.{zone.Name}.";

            var change = new Change
            {
                Additions = new[]
                {
                    new ResourceRecordSet
                    {
                        Name = recordName,
                        Type = "TXT",
                        Ttl = 60,
                        Rrdatas = values.ToArray()
                    }
                }
            };

            await _dnsService.Changes.Create(change, _credsParameters.ProjectId, zone.Id).ExecuteAsync();
        }
Let me know what you think?

**Reply by vigouredelaruse** [reply]
sweet glad you feel motivated

i haven't been typing into the fork i have going for a bit because i'm trying to get a different docker appliance out the door.

probably the quickest path forward is that if you have cycles and you grok the godaddy api then you should have write access to the fork, and we commit code till it works

all my latest changes are in the repo- one thing i just noticed i've been committing to the master branch of the fork... hmm

so would you like write access to the fork?
